snap2.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SNAP2 1"
.TH SNAP2 1 "2012-01-02" "1.00" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
snap2 \- a method for evaluating effects of single amino acid substitutions on protein function
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
snap2 [\s-1OPTION\s0]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1SNAP\s0 is a method for evaluating effects of single amino acid substitutions on protein function.
.PP
Single Nucleotide Polymorphisms (SNPs) represent a very large portion of all genetic variations. SNPs found in the coding regions of genes are often non-synonymous, changing a single amino acid in the encoded protein sequence. SNPs are either \*(L"neutral\*(R" in the sense that the resulting point-mutated protein is not functionally discernible from the wild-type, or they are \*(L"non-neutral\*(R" in that the mutant and wild-type differ in function. The ability to identify non-neutral substitutions in an ocean of SNPs could significantly aid targeting disease causing detrimental mutations, as well as SNPs that increase the fitness of particular phenotypes.
.PP
\&\s-1SNAP\s0 is a neural-network based method that uses in silico derived protein information (e.g. secondary structure, conservation, solvent accessibility, etc.) in order to make predictions regarding functionality of mutated proteins. The network takes protein sequences and lists of mutants as input, returning a score for each substitution. These scores can then be translated into binary predictions of effect (present/absent) and reliability indices (\s-1RI\s0). \s-1SNAP\s0 returns its results to the user via e\-mail.
.SS "Output format"
.IX Subsection "Output format"
Self-annotating, see example outputs in \fI/tmp/hecht3/share/snap2/example\fR.  No output for mutations below minimum reliability or expected accuracy but see \fB\-\-print\-collection\fR.
.PP
Value ranges in output:
.PP
.Vb 1
\& <mutation> [Non\-neutral|Neutral] [0\-9] [0\-100%]
.Ve
.SH "REFERENCES"
.IX Header "REFERENCES"
.IP "Bromberg, Y., Yachdav, G., and Rost, B. (2008). \s-1SNAP\s0 predicts effect of mutations on protein function. Bioinformatics, 24(20), 2397\-8." 4
.IX Item "Bromberg, Y., Yachdav, G., and Rost, B. (2008). SNAP predicts effect of mutations on protein function. Bioinformatics, 24(20), 2397-8."
.SH "OPTIONS"
.IX Header "OPTIONS"
.PD 0
.IP "\-i" 4
.IX Item "-i"
.PD
Input file in fasta format with \f(CW\*(C`.fasta\*(C'\fR extension.
.IP "\-m" 4
.IX Item "-m"
Mutant file in <from><AAnumber><to> format:
.Sp
.Vb 3
\& C30Y
\& R31W
\& G48D
.Ve
.Sp
\&\s-1OR:\s0 Keyword 'all' will mutate every position into each of the 19 non-native amino acids. I.e.: a full mutagenesis analysis
.IP "\-o" 4
.IX Item "-o"
Output file path
.IP "\-cpus" 4
.IX Item "-cpus"
Number of processors (cores) to use. Default: \fI1\fR.
.IP "\-\-debug" 4
.IX Item "--debug"
.PD 0
.IP "\-\-nodebug" 4
.IX Item "--nodebug"
.PD
Debugging messages, use twice to increase verbosity. default: \fB\-\-nodebug\fR
.IP "\-\-force\-cache\-store" 4
.IX Item "--force-cache-store"
Force snap2 to reproduce all files and store them in the cache. This is only sensible if the cache data is out of date.
.IP "\-\-help" 4
.IX Item "--help"
Display a short usage overview
.IP "\-\-quiet" 4
.IX Item "--quiet"
.PD 0
.IP "\-\-noquiet" 4
.IX Item "--noquiet"
.PD
silence progress messages, default: \fB\-\-noquiet\fR
.IP "\-\-version" 4
.IX Item "--version"
print the program version
.IP "\-\-workdir" 4
.IX Item "--workdir"
Working directory.  Use this option to force program reuse intermediate processing results.  Default: a new temporary directory that is removed automatically.
.SS "Expert options"
.IX Subsection "Expert options"
.IP "\-\-print\-collection" 4
.IX Item "--print-collection"
.PD 0
.IP "\-\-noprint\-collection" 4
.IX Item "--noprint-collection"
.PD
Print outputs of each of the snap2 predictions against the 10 models \- these are neural network outputs, [<int> <int> |]+ \- into the final output file. Default: \fB\-\-noprint\-collection\fR.
.Sp
.Vb 1
\& <mutation> => [<int> <int> |]+ sum = [\-100\-100]
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
\&\f(CW\*(C`snap2 \-i /tmp/hecht3/share/snap2/example/MT4_HUMAN.fasta \-m /tmp/hecht3/share/snap2/example/MT4_HUMAN.muts \-o MT4_HUMAN.snap2\*(C'\fR
.PP
\&\f(CW\*(C`snap2 \-i /tmp/hecht3/share/snap2/example/MT4_HUMAN.fasta \-m all \-o MT4_HUMAN.snap2\*(C'\fR
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\s-1SNAP2CONF\s0" 4
.IX Item "SNAP2CONF"
Location of snap2rc configuration file to use, overriding other configuration files
.SH "FILES"
.IX Header "FILES"
.IP "\fI/tmp/hecht3/share/snap2/snap2rc.default\fR" 4
.IX Item "/tmp/hecht3/share/snap2/snap2rc.default"
Default configuration file. See this file for a description of the parameters.
.IP "\fI/etc/snap2rc\fR" 4
.IX Item "/etc/snap2rc"
System configuration file overriding values in \fI/tmp/hecht3/share/snap2/snap2rc.default\fR
.IP "\fI~/.snap2rc\fR" 4
.IX Item "~/.snap2rc"
User configuration file overriding values in \fI/etc/snap2rc\fR
.ie n .IP "\fI\fI$SNAP2CONF\fI\fR" 4
.el .IP "\fI\f(CI$SNAP2CONF\fI\fR" 4
.IX Item "$SNAP2CONF"
If this environment variable is set \fI~/.snap2rc\fR is disregarded and the value of the variable is read for configuration options overriding \fI/etc/snap2rc\fR
.SH "AUTHOR"
.IX Header "AUTHOR"
Maximilian Hecht <hecht@rostlab.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
(C) Copyright 2012, Maximilian Hecht
.PP
\&\s-1NON\s0 \s-1COMMERCIAL\s0 \s-1SOFTWARE\s0 \s-1LICENSE\s0 \s-1AGREEMENT\s0
.PP
1. \s-1GENERAL\s0.
.PP
\&\s-1THE\s0 \s-1TERMS\s0 \s-1OF\s0 \s-1THIS\s0 \s-1LICENSE\s0 \s-1APPLY\s0 \s-1TO\s0 NON-PROFIT \s-1ORGANZIATIONS\s0 \s-1AND\s0 \s-1USERS\s0 \s-1WHO\s0 \s-1ARE\s0 \s-1EMPLOYEES\s0 \s-1OF\s0 NON-PROFIT \s-1ORGANIZAIONS\s0. \s-1FOR\s0 \s-1PROFIT\s0 \s-1ORGANIZATION\s0 \s-1AND\s0 \s-1INDIVIDIAL\s0 \s-1USERS\s0 \s-1SHOULD\s0 \s-1CONTACT\s0 info@bio\-sof.com \s-1TO\s0 \s-1OBTAIN\s0 A \s-1COMMERCIAL\s0 \s-1LICENSE\s0
.PP
2. \s-1OWNERSHIP\s0 \s-1OF\s0 \s-1LICENSED\s0 \s-1PRODUCT\s0.
.PP
(a) Licensee acknowledges and agrees that the ``Licensed Product'' and all copies thereof are Licensor's exclusive property.  Licensee has no rights with respect to the Licensed Product except as set forth in this Agreement, and, without limiting the generality of the foregoing, may not distribute, resell, sublicense, assign or transfer the ``Licensed Product'' or any portion thereof, or modify, decompile, disassemble or otherwise change the Licensed Software without Licensor's prior written consent.
.PP
(b) Upon any termination, cancellation, or expiration hereof, Licensee shall immediately delete the Licensed Software from all of its computer systems and destroy any copies of the Licensed Documentation in its possession.
.PP
3. \s-1GRANT\s0 \s-1OF\s0 \s-1LICENSE\s0.
.PP
a. Licensor hereby grants to Licensee, and Licensee hereby accepts, a personal, Non-exclusive and non-transferable license to Use the Licensed Software at the Licensed Site during the term hereof, and to use the Licensed Documentation during such term in support of the Use of the Licensed Software. Additionally individuals who are employees or contractors of Licensee who log into computers at the Licensed Site from their personal home or while traveling, are entitled to Use the Licensed Software at such remote locations, provided that such Use is undertaken for and on behalf of the Licensee and that such employees or contractors report directly to employees of the Licensee whose usual and regular place of work is the Licensed Site.
.PP
b. Other than as set forth in paragraph 3(a) above, a separate license shall be required, together with the payment of additional annual license fees and charges, to use the Licensed Software at any location other than the Licensed Site.
.PP
4. \s-1REPRODUCTION\s0 \s-1OF\s0 \s-1LICENSED\s0 \s-1PRODUCT\s0.
.PP
Licensee may reproduce the Software for Use in accordance with the terms of the limited License granted to the Licensee in Section 2.   Reproduction of the Licensed Software for any other reason is strictly prohibited.  All copies of the Software, in whole or in part, shall contain the Licensor's restrictive and proprietary notices as they appear on the copies of Software provided by Licensor.
.PP
5. \s-1LIMITATION\s0 \s-1OF\s0 \s-1LIABILITY\s0.
.PP
a. \s-1IN\s0 \s-1NO\s0 \s-1EVENT\s0 \s-1SHALL\s0 \s-1LICENSOR\s0 \s-1BE\s0 \s-1LIABLE\s0 \s-1TO\s0 \s-1LICENSEE\s0 \s-1FOR\s0 \s-1ANY\s0 \s-1INDIRECT\s0, \s-1SPECIAL\s0 \s-1OR\s0 \s-1CONSEQUENTIAL\s0 \s-1DAMAGES\s0 \s-1OR\s0 \s-1LOST\s0 \s-1PROFITS\s0, \s-1ARISING\s0 \s-1OUT\s0 \s-1OF\s0 \s-1OR\s0 \s-1RELATED\s0 \s-1TO\s0 \s-1THIS\s0 \s-1LICENSE\s0 \s-1AGREEMENT\s0 \s-1OR\s0 \s-1THE\s0 \s-1PERFORMANCE\s0 \s-1OR\s0 \s-1BREACH\s0 \s-1THEREOF\s0, \s-1EVEN\s0 \s-1IF\s0 \s-1THE\s0 \s-1LICENSOR\s0 \s-1HAS\s0 \s-1BEEN\s0 \s-1ADVISED\s0 \s-1OF\s0 \s-1THE\s0 \s-1POSSIBILITY\s0 \s-1THEREOF\s0.
.PP
b. \s-1IN\s0 \s-1NO\s0 \s-1EVENT\s0 \s-1SHALL\s0 \s-1LICENSOR\s0 \s-1BE\s0 \s-1LIABLE\s0 \s-1TO\s0 \s-1LICENSEE\s0 \s-1FOR\s0 \s-1ANY\s0 \s-1DAMAGES\s0 \s-1RESULTING\s0 \s-1FROM\s0 \s-1OR\s0 \s-1RELATED\s0 \s-1TO\s0 \s-1ANY\s0 \s-1FAILURE\s0 \s-1OF\s0 \s-1THE\s0 \s-1SOFTWARE\s0 \s-1PRODUCTS\s0, \s-1INCLUDING\s0, \s-1BUT\s0 \s-1NOT\s0 \s-1LIMITED\s0 \s-1TO\s0 \s-1LOSS\s0 \s-1OF\s0 \s-1DATA\s0, \s-1OR\s0 \s-1DELAY\s0 \s-1OF\s0 \s-1THE\s0 \s-1LICENSOR\s0 \s-1IN\s0 \s-1THE\s0 \s-1DELIVERY\s0 \s-1OF\s0 \s-1THE\s0 \s-1LICENSED\s0 \s-1PRODUCT\s0 \s-1OR\s0 \s-1IN\s0 \s-1THE\s0 \s-1PERFORMANCE\s0 \s-1OF\s0 \s-1SERVICES\s0 \s-1UNDER\s0 \s-1THIS\s0 \s-1LICENSE\s0 \s-1AGREEMENT\s0.
.PP
c. \s-1IN\s0 \s-1NO\s0 \s-1EVENT\s0 \s-1SHALL\s0 \s-1LICENSEE\s0 \s-1BE\s0 \s-1LIABLE\s0 \s-1TO\s0 \s-1LICENSOR\s0 \s-1FOR\s0 \s-1ANY\s0 \s-1LOST\s0 \s-1PROFITS\s0, \s-1LOST\s0 \s-1OPPORTUNITY\s0 \s-1COSTS\s0 \s-1OR\s0 \s-1ANY\s0 \s-1SPECIAL\s0, \s-1INDIRECT\s0, \s-1CONSEQUENTIAL\s0 \s-1OR\s0 \s-1INCIDENTAL\s0 \s-1DAMAGES\s0, \s-1HOWEVER\s0 \s-1CAUSED\s0 \s-1AND\s0 \s-1ON\s0 \s-1ANY\s0 \s-1THEORY\s0 \s-1OF\s0 \s-1LIABILITY\s0.


ChangeLog
snapfun (1.0.0) UNRELEASED; urgency=low

  * lkajan: updated license file and man page

 -- Laszlo Kajan <lkajan@rostlab.org>  Tue, 03 Jan 2012 18:05:06 +0100


snap2.pl
#!/usr/bin/perl -w
use strict;
use feature qw(say);
use Getopt::Long;
use Cwd 'abs_path';
use File::Temp qw||;
use File::chdir;
use File::Basename;
use File::Path qw(make_path);
use Carp qw(cluck :DEFAULT);

our ($allmuts,$config,$sequence,@mutants,$name,@sequence_array,$swissdb,$db_swiss,$swiss_dat,$big80,$phat_matrix);
my ($snap2dir,$fannlib,$cpu);
BEGIN {
    use Config::IniFiles;
    #our $VERSION = "__VERSION__";

    #my ( $defaultconfig, $etcconfig );
    #if( -e "__pkgdatadir__/snapfunrc.default" ) { $defaultconfig = Config::IniFiles->new( -file => "__pkgdatadir__/snapfunrc.default" ); }
    #if( -e "__sysconfdir__/snapfunrc" ) { $etcconfig = Config::IniFiles->new( -file => "__sysconfdir__/snapfunrc", -import => $defaultconfig ); } else { $etcconfig = $defaultconfig; }
    #if( ( $ENV{SNAPFUNCONF} && -e "$ENV{SNAPFUNCONF}" ) || -e "$ENV{HOME}/.snapfunrc" ) { $config = Config::IniFiles->new( -file => $ENV{SNAPFUNCONF} || "$ENV{HOME}/.snapfunrc", -import => $etcconfig ); } else { $config = $etcconfig; }

    $config=Config::IniFiles->new( -file => "/mnt/project/resnap/trunk/snap2rc.default");
    $snap2dir = $config->val('snap2', 'snap2dir');
    $swissdb=$config->val('blast','swiss');
    $swiss_dat=$config->val('data','swiss_dat');
    $db_swiss=$config->val('data','db_swiss');
    $big80=$config->val('blast','big80');
    $phat_matrix=$config->val('data','phat_matrix');
    $fannlib=$config->val('snap2','fann_lib');
    $cpu=$config->val('snap2', 'blastpgp_processors');
}

use lib glob( "$snap2dir/lib" );
use lib glob( $fannlib );
use Run;
use Extract;
use Predict;
use Prediction;

my($in,$out,$mut,$workdir,$plot,$labeled_muts_file,$only_extract,@labels,$pc,$quiet,$fcs);
my $debug=0;
$cpu=1;
my $args_ok=GetOptions( 'in=s'    =>  \$in,
                        'out=s'    =>  \$out,
                        'workdir=s'    =>  \$workdir,
                        'debug+'     => \$debug,
                        'muts=s'    =>  \$mut,
                        'label=s'   => \$labeled_muts_file,
                        'extract'   =>\$only_extract,
                        'print-collections' => \$pc,
                        'quiet' => \$quiet,
                        'force-cache-store' => \$fcs,
                        'cpus=i' => \$cpu,
                        'plot'  => \$plot
);
sub die_usage{
    my $msg=shift;
    say "\nDESCRIPTION:\nPredict functional effect of non-synonymous single nucleotide polymorphisms";
    say "\nUSAGE:\nsnap2 -i <input fasta> -o <output file> -m <mutations file> [OPTIONS]";
    say "\nMANDATORY:";
    say "-i, --in <file>\n\tInput sequence in fasta format";
    say "\n-m, --mut <file>\n\tMutation file in format: [A-Z][0-9]+[A-Z] e.g: 'A32S'. One per line";
    say "\n-o, --out <file>\n\tOutput file";
    say "\nOPTIONS:";
    say "-w, --workdir <directory>\n\tWorking directory. Intermediate files will be saved here";
    say "\n-l, --labels <file>\n\tMutation file with labels in format: [A-Z][0-9]+[A-Z]".'\s'."[0|1] e.g: 'A32S 1' (where 1 = non-neutral, 0 = neutral). One per line";
    say "\n-e, --extract\n\tOnly extract feature values. Will be saved to <out file>.features";
    say "\n-p, --print-collections\n\tAlso print the raw outputs for each network";
    say "\n-q, --quiet\n\tSilence announcements";
    say "\n-d, --debug\n\tPrint debugging messages. Use twice to increase level of debugging messages";
    say "\n-c, --cpus\n\tNumber of cpu cores";
    die "\n$msg\n";
}
die_usage("No input file defined") unless $in;
die_usage("No output file defined") unless $out;
die_usage("Unknown argument") unless $args_ok;

if ($workdir){
    $workdir=Cwd::realpath( $workdir );
}
else {
    $workdir=File::Temp::tempdir( CLEANUP=> !$debug);
}

warn "Working directory: $workdir\n" unless $quiet;

$in=Cwd::realpath($in);
$out=Cwd::realpath($out);
$name=fileparse($out,qr/\.[^.]*/);

warn "Job name: $name\n" unless $quiet;

#Read sequence file
open (FHIN,$in) || confess "\nError: unable to open sequence file: $in\n" ;
while (<FHIN>){
    next if /^>/o;
    chomp($_);
    $sequence.=$_;
}
$sequence=~s/\s//g;
close FHIN;

#sequence as array
@sequence_array=split(//o,$sequence);

#Read mutants file
if ($labeled_muts_file){
    $labeled_muts_file=Cwd::realpath($labeled_muts_file);
    #Get the corresponding class label if available
    open (LABEL,$labeled_muts_file)|| confess "\nError: unable to open labeled mutation file: $labeled_muts_file\n";
    while (<LABEL>){
        confess "\nError: Invalid mutation line: $_\n" if !/^[A-Z]\d+[A-Z]\s+\d$/o;
        my ($mut,$label)=split;
        push (@mutants,$mut);
        push (@labels,$label);
    }
    close LABEL;

    #write mutant file in correct format (will be passed on to reprof and sift)
    $mut="$workdir/$name.muts";
    open MUT,">$mut" || confess "\nError: unable to write $mut\n";
    print MUT join "\n",@mutants;
    close MUT;
}
elsif ($mut){
    if ($mut eq 'all'){
        @mutants=allmuts(\$mut,$workdir,\@sequence_array,$debug); 
        $allmuts=1;
        $mut="$workdir/$name.allmuts";
    }
    else{
        $mut=Cwd::realpath($mut);
        open (FHIN,$mut) || confess "\nError: unable to open mutations file: $mut\n";
        while(<FHIN>){
            #check every mutant
            confess "\nError: Invalid mutation line: $_\n" unless /^[ARNDCQEGHILKMFPSTWYV]\d+[ARNDCQEGHILKMFPSTWYV]$/o;
            chomp $_;
            push (@mutants,$_);
        }
        close FHIN;
    }
}
else {
    die_usage("No mutation file defined");
}

#Run all external programs
warn "\nRunning external programs..\n" unless $quiet;
Run::all($mut,$workdir,$fcs,$cpu,$allmuts,$debug);

#Extract feature values
warn "\nExtracting feature values..\n" unless $quiet;
my $data=Extract::all($workdir,$debug>1 ? 1 : 0);

#write out the featurefile (for training purpose, won't be used anymore afterwards)
confess "Number of labels is not consistent with number of data points" if ($labeled_muts_file && scalar(@labels) != scalar(@$data));
my $extract="$workdir/$name.features";
open FEATURES,">$extract" || confess "\nError: unable to write $extract\n";
foreach my $i (0..@{$data}-1) {
    say FEATURES join(" ",@{@$data[$i]});
    say FEATURES ($labels[$i]==1 ? "0 1" : "1 0") if ($labeled_muts_file);
}
close FEATURES;

exit(0) if $only_extract;

#Run all the neural networks
warn "\nRunning predictions..\n" unless $quiet;
my @predictions=Predict::all($data,"$snap2dir/models",$debug>1 ? 1 : 0);
confess "Number of predictions is not consistent with number of data points" if scalar(@predictions) != scalar(@$data);

#Expected accuracy as obtained from 10-fold cross-validation
my %expected_accuracy=( -9 => '97%',
                        -8 => '93%',
                        -7 => '87%',
                        -6 => '82%',
                        -5 => '78%',
                        -4 => '72%',
                        -3 => '66%',
                        -2 => '61%',
                        -1 => '57%',
                         0 => '53%',
                         1 => '59%',
                         2 => '63%',
                         3 => '66%',
                         4 => '71%',
                         5 => '75%',
                         6 => '80%',
                         7 => '85%',
                         8 => '91%',
                         9 => '95%');

#Write output file
my $prediction=new Prediction();
open OUT,">$out" or confess "Unable to write output file: $out";
foreach my $data_point (0..@predictions-1) {
    my ($neu,$non)=qw(0 0);
    print OUT $mutants[$data_point] . " => " if $pc;
    foreach my $network (@{$predictions[$data_point]}) {
        $neu+=$$network[0];
        $non+=$$network[1];
        print OUT int(100*$$network[0]) ." ". int(100*$$network[1]) ."\t| " if $pc;
    }
    say OUT "Sum = ". int(100*($non-$neu)/scalar(@{$predictions[0]})) if $pc;
    if ($labeled_muts_file){
        $prediction->add([$neu/10,$non/10],($labels[$data_point] == 1 ? [0,1] : [1,0]));
    }
    my $ri=$non-$neu;
    say OUT $mutants[$data_point] . " => Prediction: " . ($ri>0 ? "Non-neutral" : "Neutral") . "\tReliability Index: " . int(abs($ri)) . "\tExpected accuracy: " . $expected_accuracy{int($ri)}; 
    
}
close OUT;
$prediction->write("$out.labeledpred") if $labeled_muts_file;
if ($plot){
    use lib glob ("$snap2dir/plots");
    use Plot;
    Plot::from_prediction($name,\@sequence_array,\@predictions,$out,$debug);
}
warn "\nOutput written to $out\n" unless $quiet;
exit(0);

#--------------------------------------------------
# Sub-routines 
#-------------------------------------------------- 

sub allmuts{
    my ($mut,$workdir,$seq_arr,$debug)=@_;
    my @mutants;
    my @amino_acids=qw( A R N D C Q E G H I L K M F P S T W Y V );
    for (my $i = 0; $i < scalar(@$seq_arr); $i++) {
        my $wt=($$seq_arr[$i]!~/[ARNDCQEGHILKMFPSTWYV]/o ? "A" : $$seq_arr[$i]);
        foreach my $aa (@amino_acids) {
            push @mutants,$wt . ($i+1) . $aa unless $wt eq $aa;
        }
    }
    open MUT,">$workdir/$name.allmuts" or confess "Unable to write mutant file: $workdir/$name.allmuts";
    say MUT join "\n",@mutants;
    close MUT;

    return @mutants;
}
